### ISSUE TITLE
using BTreeMaps in semantic package instead of HashMap results in valgrind errors

### ISSUE THREAD

**Issue by adrian-thurston** [influxdata/flux#2636]
On the [fix/semantic-btree-map](https://github.com/influxdata/flux/tree/fix/semantic-btree-map) valgrind reports an invalid free. This branch switches all HashMaps to BTreeMaps in order to achieve a stable serialization of the environment. 

Uncovered while working on:
https://github.com/influxdata/flux/issues/2585

```
cd libflux && cargo build
   Compiling flux v0.3.0 (/home/thurston/devel/flux/libflux/src/flux)
   Compiling libstd v0.1.0 (/home/thurston/devel/flux/libflux/src/libstd)
    Finished dev [unoptimized + debuginfo] target(s) in 16.59s
cc -g -Wall -Werror libflux/c/*.c -I./libflux/include -L./libflux/target/debug -lflux -llibstd -o libflux/c/libflux_memory_tester
LD_LIBRARY_PATH=/home/thurston/devel/flux/libflux/target/debug valgrind --leak-check=full --error-exitcode=1 --num-callers=100 libflux/c/libflux_memory_tester
==15089== Memcheck, a memory error detector
==15089== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==15089== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==15089== Command: libflux/c/libflux_memory_tester
==15089==
Testing AST functions...
Parsing to AST
Marshaling to JSON
  json buffer is length 899
Marshaling to FlatBuffer
  FlatBuffer is length 760

Testing semantic graph functions...
Parsing to AST
Analyzing (expect success)
Marshaling to FlatBuffer
  FlatBuffer is length 880
==15089== Invalid free() / delete / delete[] / realloc()
==15089==    at 0x4C30D3B: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15089==    by 0x4F2CB92: alloc::alloc::dealloc (alloc.rs:106)
==15089==    by 0x4F2C979: <alloc::alloc::Global as core::alloc::Alloc>::dealloc (alloc.rs:177)
==15089==    by 0x4F8C20E: alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned,K,V,alloc::collections::btree::node::marker::Leaf>::deallocate_and_ascend (node.rs:504)
==15089==    by 0x5093FF0: <alloc::collections::btree::map::IntoIter<K,V> as core::ops::drop::Drop>::drop (map.rs:1317)
==15089==    by 0x511642E: core::ptr::real_drop_in_place (mod.rs:175)
==15089==    by 0x4FEBCD6: core::mem::drop (mod.rs:706)
==15089==    by 0x5093CB1: <alloc::collections::btree::map::BTreeMap<K,V> as core::ops::drop::Drop>::drop (map.rs:132)
==15089==    by 0x51176FE: core::ptr::real_drop_in_place (mod.rs:175)
==15089==    by 0x511AB0E: core::ptr::real_drop_in_place (mod.rs:175)
==15089==    by 0x5119B63: core::ptr::real_drop_in_place (mod.rs:175)
==15089==    by 0x5119774: core::ptr::real_drop_in_place (mod.rs:175)
==15089==    by 0x5114D57: core::ptr::real_drop_in_place (mod.rs:175)
==15089==    by 0x5083EC1: drop_in_place<[flux::semantic::nodes::Statement]> (mod.rs:165)
==15089==    by 0x5083EC1: <alloc::vec::Vec<T> as core::ops::drop::Drop>::drop (vec.rs:2237)
==15089==    by 0x5111B24: core::ptr::real_drop_in_place (mod.rs:175)
==15089==    by 0x5116181: core::ptr::real_drop_in_place (mod.rs:175)
==15089==    by 0x511A1FB: core::ptr::real_drop_in_place (mod.rs:175)
==15089==    by 0x5083F91: drop_in_place<[flux::semantic::nodes::File]> (mod.rs:165)
==15089==    by 0x5083F91: <alloc::vec::Vec<T> as core::ops::drop::Drop>::drop (vec.rs:2237)
==15089==    by 0x5111464: core::ptr::real_drop_in_place (mod.rs:175)
==15089==    by 0x511001D: core::ptr::real_drop_in_place (mod.rs:175)
==15089==    by 0x5117173: core::ptr::real_drop_in_place (mod.rs:175)
==15089==    by 0x510F9AD: core::ptr::real_drop_in_place (mod.rs:175)
==15089==    by 0x4F1456D: flux_free_semantic_pkg (lib.rs:215)
==15089==    by 0x108F40: test_semantic (main.c:65)
==15089==    by 0x108C7C: main (main.c:14)
==15089==  Address 0x5e409e8 is 0 bytes inside data symbol "_ZN5alloc11collections5btree4node15EMPTY_ROOT_NODE17h2eaf60d0702cf6a2E"
==15089==
Parsing to AST
Analyzing (expect failure)
  error: cannot unify int with float

Testing semantic analyzer...
Parsing and analyzing "x = 10"
Parsing and analyzing "y = x * x"
Parsing and analyzing "z = a + y" (expect failure)
  error: undeclared variable a

Testing flux_get_env_stdlib
  got a buffer of size 85360

==15089==
==15089== HEAP SUMMARY:
==15089==     in use at exit: 0 bytes in 0 blocks
==15089==   total heap usage: 24,855 allocs, 24,858 frees, 2,604,643 bytes allocated
==15089==
==15089== All heap blocks were freed -- no leaks are possible
==15089==
==15089== For counts of detected and suppressed errors, rerun with: -v
==15089== ERROR SUMMARY: 3 errors from 1 contexts (suppressed: 0 from 0)
Makefile:141: recipe for target 'test-valgrind' failed
make: *** [test-valgrind] Error 1
```

**Comment by wolffcm** [comment]
If we can reproduce this without using any unsafe code, then it's a bug in the Rust compiler (or possible valgrind, but that seems less likely), I think.  So a next step might be to: 
- Write a Rust unit test with no unsafe code that creates a semantic graph with at least one assignment, and let it be dropped
- Run that unit test in valgrind

If there is no invalid free, then the problem is on our side for sure.

**Comment by fchikwekwe** [comment]
I was able to replicate the error that @adrian-thurston mentioned above by running `make test-valgrind` in a linux docker container. 

I wrote the following Rust unit test in `libflux/libstd` and ran it in valgrind.

```    
#[test]
    fn sem_pkg_valgrind() {
        let ast: ast::Package = flux::parser::parse_string("test", "package foo\nx = 1 + 1").into();
        let pkg = match analyze(ast) {
            Ok(sem_pkg) => {
                Box::new(sem_pkg)
            }
            Err(e) => panic!("{}", e)
        };

        let _sem_pkg = match flatbuffers::serialize(pkg.as_ref()) {
            Ok(fb) => fb,
            Err(e) => panic!(e)
        };
    }
```

There was a memory error with the following output. This doesn't seem to be super related to what Adrian is working on above, but it is additional information.

```
builder@c3dfe819c310:/flux/libflux/target/debug$ valgrind ./libstd-c770529cf7679b89 sem_pkg_valgrind
==5953== Memcheck, a memory error detector
==5953== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==5953== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==5953== Command: ./libstd-c770529cf7679b89 sem_pkg_valgrind
==5953== 
==5953== Syscall param statx(file_name) points to unaddressable byte(s)
==5953==    at 0x4986F59: syscall (syscall.S:38)
==5953==    by 0x5CDA5B: statx (weak.rs:90)
==5953==    by 0x5CDA5B: std::sys::unix::fs::try_statx (fs.rs:132)
==5953==    by 0x5C892E: std::sys::unix::fs::stat (fs.rs:995)
==5953==    by 0x1C42C4: metadata<&std::path::PathBuf> (fs.rs:1591)
==5953==    by 0x1C42C4: term::terminfo::searcher::get_dbpath_for_term (searcher.rs:50)
==5953==    by 0x1C35C2: term::terminfo::TermInfo::from_name (mod.rs:87)
==5953==    by 0x1C33D0: term::terminfo::TermInfo::from_env (mod.rs:73)
==5953==    by 0x1CE019: new<std::io::stdio::Stdout> (mod.rs:232)
==5953==    by 0x1CE019: term::stdout (lib.rs:61)
==5953==    by 0x1AEBDB: test::console::run_tests_console (console.rs:264)
==5953==    by 0x1B65F3: test::test_main (lib.rs:121)
==5953==    by 0x1B7600: test::test_main_static (lib.rs:140)
==5953==    by 0x192915: libstd::main (in /flux/libflux/target/debug/libstd-c770529cf7679b89)
==5953==    by 0x1892AF: std::rt::lang_start::{{closure}} (rt.rs:67)
==5953==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==5953== 
==5953== Syscall param statx(buf) points to unaddressable byte(s)
==5953==    at 0x4986F59: syscall (syscall.S:38)
==5953==    by 0x5CDA5B: statx (weak.rs:90)
==5953==    by 0x5CDA5B: std::sys::unix::fs::try_statx (fs.rs:132)
==5953==    by 0x5C892E: std::sys::unix::fs::stat (fs.rs:995)
==5953==    by 0x1C42C4: metadata<&std::path::PathBuf> (fs.rs:1591)
==5953==    by 0x1C42C4: term::terminfo::searcher::get_dbpath_for_term (searcher.rs:50)
==5953==    by 0x1C35C2: term::terminfo::TermInfo::from_name (mod.rs:87)
==5953==    by 0x1C33D0: term::terminfo::TermInfo::from_env (mod.rs:73)
==5953==    by 0x1CE019: new<std::io::stdio::Stdout> (mod.rs:232)
==5953==    by 0x1CE019: term::stdout (lib.rs:61)
==5953==    by 0x1AEBDB: test::console::run_tests_console (console.rs:264)
==5953==    by 0x1B65F3: test::test_main (lib.rs:121)
==5953==    by 0x1B7600: test::test_main_static (lib.rs:140)
==5953==    by 0x192915: libstd::main (in /flux/libflux/target/debug/libstd-c770529cf7679b89)
==5953==    by 0x1892AF: std::rt::lang_start::{{closure}} (rt.rs:67)
==5953==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==5953== 

running 1 test
test tests::sem_pkg_valgrind ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 3 filtered out

==5953== 
==5953== HEAP SUMMARY:
==5953==     in use at exit: 0 bytes in 0 blocks
==5953==   total heap usage: 5,660 allocs, 5,660 frees, 530,136 bytes allocated
==5953== 
==5953== All heap blocks were freed -- no leaks are possible
==5953== 
==5953== For counts of detected and suppressed errors, rerun with: -v
==5953== ERROR SUMMARY: 2 errors from 2 contexts (suppressed: 0 from 0)
```

**Comment by adrian-thurston** [comment]
Thank you @fchikwekwe! So I think this rules out a pure rust issue with the code and a rust compiler bug and points to an issue with the C/rust interface. Something in the tree that is exported to C is not exported cleanly and when it goes back into rust to be freed there is a failure.

**Comment by adrian-thurston** [comment]
It looks like passing any *empty* `BTreeMap` from Rust to C, then passing it back to Rust to be freed results in the above `valgrind` error.

As @wolffcm suggested [on slack](https://influxdata.slack.com/archives/C0AD7FFS9/p1584376333032400), it is caused by the new `BTreeMap` in `VariableAssgn` struct in the `semantic` package, but only when empty.

Boxing or making the field optional doesn't help.

There could be some kind of optimization in `BTreeMap` for the empty case that messes with the normal memory allocation rules.

It's possible it can be safely ignored but this would take some verification.

**Comment by adrian-thurston** [comment]
Shared empty root looks like the problem. From `map.rs` in `BTreeMap`:

```
    pub fn new() -> BTreeMap<K, V> {
        BTreeMap { root: node::Root::shared_empty_root(), length: 0 }
    }

**Comment by adrian-thurston** [comment]
The shared empty root is a statically allocated node and rust uses a comparison of its address to determine if the root is this static node.

I don't know why when we call back into rust to free the address is different. I would think they would be the same. This itself might be worth investigating.

Also would be great to understand why there are two casts.

```
impl<K, V> NodeHeader<K, V> {
    fn is_shared_root(&self) -> bool {
        ptr::eq(self, &EMPTY_ROOT_NODE as *const _ as *const _)
    }
}
```

**Comment by adrian-thurston** [comment]
Also relevant.
https://github.com/rust-lang/rust/issues/50266

**Comment by adrian-thurston** [comment]
Problem stems from allocating in `libstd` and freeing in `libflux`.

Normally when linking with Go programs we use the static archives and the linker creates a single data segment for both pieces of code, leading to a valid comparison of the static root in the `BTreeMap` deallocator. Static linking is why none of the non-valgrind test cases failed a `SIGABRT` after switching to `BTreeMap`

The valgrind test is linked dynamically, and each library gets its own data segment, leading to two static root nodes. The deallocator cannot detect that it has a pointer to the static root node and attempts to free it.

Fix is to always free in the same library we allocated in.

*** Possibly we want to also consider treating data that came out of one lib strictly incompatible with other Rust libs. If Rust code must share heap data, we should put the code in the same lib. It is possible there are other gotchas that could arise from invalid comparisons, originating from static data. For example, something containing a map might fail to compare properly. I have not proven this, but it's a concern I have now.

To summarize, the conditions necessary to reproduce:

1. Allocate in one rust library, free in another.
2. An *empty* `BTreeMap` is included in the data passed between
3. Dynamic linking the libs into the containing program.

**Comment by wolffcm** [comment]
Oh, wow, it makes perfect sense now.

We originally separated our rust code into `libflux` and `liblibstd` in order to maintain a strict separation between the core of Flux, i.e., its syntax and typing system, and code that is aware of the definitions in the standard library.  I still think this separation is good, but clearly we need to be more careful.

The issue here is related to allocating a semantic graph in `liblibstd` and freeing it in `libflux`, but we also do the reverse for the AST: we parse in `libflux` and then we consume and free the AST in `liblibstd` when we call `flux_analyze` to produce the semantic graph.  We just got lucky that the AST doesn't run afoul of the `BTreeMap` issue.

The easiest solution to me (at least in the short term) seems to be to put all of our public-facing API into `liblibstd`.

I had used dynamic linking for the valgrind CI job because it was easier to get it working, but with some hindsight here I think it's a good thing if it lets us find issues like this. I don't think we want the limitation that our Rust code is only used in statically linked binaries.

**Comment by adrian-thurston** [comment]
Yes it was great you did that. I think we should be expanding our use of valgrind, both static and dynamic, and with other build parameters modified.

**Comment by adrian-thurston** [comment]
Now static linking flux into the valgrind test so it matches the how we link the using CGO. With static linking we don't experience this kind of instability. The root cause will be fixed with future commits so we can safely bring back and test dynamic linking.
